File: Tests/ReminderCoreTests/ReminderEngineTests.swift
import XCTest
@testable import ReminderCore

/// Records what the engine asked to present, so tests can assert on user-visible
/// behaviour rather than internal state alone.
final class RecordingPresenter: ReminderPresenting {
    var presented: [Reminder] = []
    var dismissAllCount = 0

    func present(_ reminder: Reminder, settings: Settings) {
        presented.append(reminder)
    }

    func dismissAll() {
        dismissAllCount += 1
    }
}

@MainActor
final class ReminderEngineTests: XCTestCase {

    private var calendar: Calendar = {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        return cal
    }()

    private func date(
        _ year: Int, _ month: Int, _ day: Int, _ hour: Int = 0, _ minute: Int = 0
    ) -> Date {
        var comps = DateComponents()
        comps.year = year; comps.month = month; comps.day = day
        comps.hour = hour; comps.minute = minute; comps.second = 0
        return calendar.date(from: comps)!
    }

    /// Builds an engine with a controllable clock and no reminders.
    private func makeEngine(
        reminders: [Reminder] = [],
        settings: Settings = Settings(),
        now: Date
    ) -> (ReminderEngine, MutableDateProvider, RecordingPresenter, InMemoryDataStore) {
        let store = InMemoryDataStore(
            data: AppData(reminders: reminders, settings: settings, events: [])
        )
        let clock = MutableDateProvider(now: now)
        let presenter = RecordingPresenter()
        let engine = ReminderEngine(
            store: store, dateProvider: clock, presenter: presenter, calendar: calendar
        )
        return (engine, clock, presenter, store)
    }

    // MARK: - Firing

    func testTickFiresDueReminderAndPresentsIt() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Weight Shift", schedule: .interval(minutes: 20),
            priority: .subtle, createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, presenter, _) = makeEngine(reminders: [reminder], now: start)

        XCTAssertTrue(engine.tick().isEmpty, "Nothing is due yet")
        XCTAssertTrue(presenter.presented.isEmpty)

        clock.advance(by: 20 * 60)
        let fired = engine.tick()
        XCTAssertEqual(fired.count, 1)
        XCTAssertEqual(presenter.presented.map(\.title), ["Weight Shift"])
    }

    /// The critical bug this guards: repeated ticks in the same due window must
    /// not re-fire and spam the user.
    func testRepeatedTicksDoNotRefireSameWindow() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Tilt Back", schedule: .interval(minutes: 60), createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, presenter, _) = makeEngine(reminders: [reminder], now: start)

        clock.advance(by: 60 * 60)
        XCTAssertEqual(engine.tick().count, 1)
        // Several more ticks a few seconds apart.
        for _ in 0..<5 {
            clock.advance(by: 5)
            XCTAssertTrue(engine.tick().isEmpty)
        }
        XCTAssertEqual(presenter.presented.count, 1)
    }

    func testHigherPriorityPresentedLastSoItSitsInFront() {
        let start = date(2026, 3, 10, 9, 0)
        var subtle = Reminder(
            title: "Shift", schedule: .interval(minutes: 20),
            priority: .subtle, createdAt: start
        )
        subtle.lastFiredAt = start
        var critical = Reminder(
            title: "Tilt", schedule: .interval(minutes: 20),
            priority: .critical, createdAt: start
        )
        critical.lastFiredAt = start

        let (engine, clock, presenter, _) = makeEngine(
            reminders: [subtle, critical], now: start
        )
        clock.advance(by: 20 * 60)
        engine.tick()

        XCTAssertEqual(presenter.presented.map(\.title), ["Shift", "Tilt"])
    }

    func testDisabledReminderDoesNotFire() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Off", schedule: .interval(minutes: 10), createdAt: start
        )
        reminder.isEnabled = false
        reminder.lastFiredAt = start
        let (engine, clock, presenter, _) = makeEngine(reminders: [reminder], now: start)

        clock.advance(by: 60 * 60)
        XCTAssertTrue(engine.tick().isEmpty)
        XCTAssertTrue(presenter.presented.isEmpty)
    }

    // MARK: - Pause

    func testPausedEngineFiresNothing() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Tilt", schedule: .interval(minutes: 20), createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, presenter, _) = makeEngine(reminders: [reminder], now: start)

        engine.setPaused(true)
        clock.advance(by: 60 * 60)
        XCTAssertTrue(engine.tick().isEmpty)
        XCTAssertTrue(presenter.presented.isEmpty)
        XCTAssertEqual(presenter.dismissAllCount, 1, "Pausing clears anything on screen")
    }

    func testTimedPauseExpiresAndResumesFiring() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Tilt", schedule: .interval(minutes: 20), createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, presenter, _) = makeEngine(reminders: [reminder], now: start)

        engine.pause(forMinutes: 30)
        clock.advance(by: 20 * 60)
        XCTAssertTrue(engine.tick().isEmpty, "Still paused")

        clock.advance(by: 15 * 60) // 35 min total, pause has expired
        XCTAssertEqual(engine.tick().count, 1)
        XCTAssertFalse(engine.settings.isPaused, "Pause auto-clears on expiry")
        XCTAssertNil(engine.settings.pausedUntil)
        XCTAssertEqual(presenter.presented.count, 1)
    }

    /// Coming back from a long pause should not dump a backlog on the user.
    func testResumeReanchorsIntervalRemindersSoNoBacklogFires() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Shift", schedule: .interval(minutes: 20), createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, presenter, _) = makeEngine(reminders: [reminder], now: start)

        engine.setPaused(true)
        clock.advance(by: 4 * 60 * 60) // paused for 4 hours
        engine.resume()

        XCTAssertTrue(engine.tick().isEmpty, "Nothing should fire the instant we resume")
        clock.advance(by: 20 * 60)
        XCTAssertEqual(engine.tick().count, 1, "Next fire is a full interval after resume")
        XCTAssertEqual(presenter.presented.count, 1)
    }

    // MARK: - Quiet hours

    func testQuietHoursSuppressNonCriticalButAllowCritical() {
        let night = date(2026, 3, 10, 23, 0)
        var settings = Settings()
        settings.quietHours = QuietHours(
            isEnabled: true, startHour: 22, startMinute: 0,
            endHour: 7, endMinute: 0, allowsCritical: true
        )
        var normal = Reminder(
            title: "Water", schedule: .interval(minutes: 20),
            priority: .normal, createdAt: night
        )
        normal.lastFiredAt = night
        var critical = Reminder(
            title: "Tilt", schedule: .interval(minutes: 20),
            priority: .critical, createdAt: night
        )
        critical.lastFiredAt = night

        let (engine, clock, presenter, _) = makeEngine(
            reminders: [normal, critical], settings: settings, now: night
        )
        clock.advance(by: 20 * 60)
        let fired = engine.tick()

        XCTAssertEqual(fired.map(\.title), ["Tilt"])
        XCTAssertEqual(presenter.presented.map(\.title), ["Tilt"])
    }

    // MARK: - User responses

    func testCompleteResetsTheIntervalAndRecordsHistory() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Tilt", schedule: .interval(minutes: 60), createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, _, _) = makeEngine(reminders: [reminder], now: start)

        clock.advance(by: 60 * 60)
        engine.tick()
        clock.advance(by: 5 * 60)
        engine.complete(id: reminder.id)

        let stored = engine.reminder(withID: reminder.id)
        XCTAssertEqual(stored?.lastFiredAt, clock.now)
        XCTAssertEqual(stored?.lastAcknowledgedAt, clock.now)
        XCTAssertEqual(
            engine.events.map(\.outcome), [.fired, .completed]
        )

        // The next fire is a full hour after completion, not after the original fire.
        clock.advance(by: 58 * 60)
        XCTAssertTrue(engine.tick().isEmpty)
        clock.advance(by: 3 * 60)
        XCTAssertEqual(engine.tick().count, 1)
    }

    func testSnoozeDelaysNextFireByConfiguredMinutes() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Tilt", schedule: .interval(minutes: 60), createdAt: start
        )
        reminder.lastFiredAt = start
        var settings = Settings()
        settings.snoozeMinutes = 10
        let (engine, clock, presenter, _) = makeEngine(
            reminders: [reminder], settings: settings, now: start
        )

        clock.advance(by: 60 * 60)
        engine.tick()
        engine.snooze(id: reminder.id)

        clock.advance(by: 9 * 60)
        XCTAssertTrue(engine.tick().isEmpty, "Still snoozed")
        clock.advance(by: 2 * 60)
        XCTAssertEqual(engine.tick().count, 1, "Fires again after the snooze")
        XCTAssertEqual(presenter.presented.count, 2)
    }

    func testSnoozeWithExplicitMinutesOverridesSetting() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Tilt", schedule: .interval(minutes: 60), createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, _, _) = makeEngine(reminders: [reminder], now: start)

        clock.advance(by: 60 * 60)
        engine.tick()
        engine.snooze(id: reminder.id, minutes: 2)

        clock.advance(by: 60)
        XCTAssertTrue(engine.tick().isEmpty)
        clock.advance(by: 90)
        XCTAssertEqual(engine.tick().count, 1)
    }

    func testDismissRecordsHistoryWithoutResettingSchedule() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Water", schedule: .interval(minutes: 60), createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, _, _) = makeEngine(reminders: [reminder], now: start)

        clock.advance(by: 60 * 60)
        engine.tick()
        let firedAt = engine.reminder(withID: reminder.id)?.lastFiredAt
        clock.advance(by: 60)
        engine.dismiss(id: reminder.id)

        XCTAssertEqual(engine.reminder(withID: reminder.id)?.lastFiredAt, firedAt)
        XCTAssertEqual(engine.events.map(\.outcome), [.fired, .dismissed])
    }

    // MARK: - CRUD

    func testAddedReminderDoesNotFireImmediately() {
        let start = date(2026, 3, 10, 9, 0)
        let (engine, clock, _, _) = makeEngine(now: start)

        engine.add(
            Reminder(
                title: "New", schedule: .interval(minutes: 15),
                createdAt: date(2020, 1, 1) // deliberately stale
            )
        )
        XCTAssertTrue(engine.tick().isEmpty, "Add re-anchors creation to now")

        clock.advance(by: 15 * 60)
        XCTAssertEqual(engine.tick().count, 1)
    }

    func testReenablingReminderReanchorsSoItDoesNotFireInstantly() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Shift", schedule: .interval(minutes: 20), createdAt: start
        )
        reminder.isEnabled = false
        reminder.lastFiredAt = date(2020, 1, 1)
        let (engine, clock, _, _) = makeEngine(reminders: [reminder], now: start)

        engine.setEnabled(true, for: reminder.id)
        XCTAssertTrue(engine.tick().isEmpty, "Should not fire from a stale timestamp")

        clock.advance(by: 20 * 60)
        XCTAssertEqual(engine.tick().count, 1)
    }

    func testDeleteRemovesReminder() {
        let start = date(2026, 3, 10, 9, 0)
        let reminder = Reminder(
            title: "Gone", schedule: .interval(minutes: 10), createdAt: start
        )
        let (engine, clock, _, _) = makeEngine(reminders: [reminder], now: start)

        engine.delete(id: reminder.id)
        XCTAssertTrue(engine.reminders.isEmpty)
        clock.advance(by: 60 * 60)
        XCTAssertTrue(engine.tick().isEmpty)
    }

    func testUpdateChangesScheduleInPlace() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Shift", schedule: .interval(minutes: 60), createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, _, _) = makeEngine(reminders: [reminder], now: start)

        reminder.schedule = .interval(minutes: 10)
        engine.update(reminder)

        clock.advance(by: 10 * 60)
        XCTAssertEqual(engine.tick().count, 1)
    }

    // MARK: - Persistence

    func testChangesArePersistedToStore() {
        let start = date(2026, 3, 10, 9, 0)
        let (engine, _, _, store) = makeEngine(now: start)

        engine.add(Reminder(title: "Persisted", schedule: .interval(minutes: 30)))

        XCTAssertEqual(store.data.reminders.map(\.title), ["Persisted"])
        XCTAssertGreaterThan(store.saveCount, 0)
    }

    func testEngineLoadsExistingDataFromStore() {
        let start = date(2026, 3, 10, 9, 0)
        let existing = Reminder(title: "Loaded", schedule: .interval(minutes: 45))
        let (engine, _, _, _) = makeEngine(reminders: [existing], now: start)
        XCTAssertEqual(engine.reminders.map(\.title), ["Loaded"])
    }

    // MARK: - History and stats

    func testHistoryIsCappedToAvoidUnboundedGrowth() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Chatty", schedule: .interval(minutes: 1), createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, _, _) = makeEngine(reminders: [reminder], now: start)

        // Fire well past the cap.
        for _ in 0..<(ReminderEngine.maxStoredEvents + 50) {
            clock.advance(by: 60)
            engine.tick()
        }
        XCTAssertLessThanOrEqual(engine.events.count, ReminderEngine.maxStoredEvents)
    }

    func testAdherenceReflectsCompletionRate() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Tilt", schedule: .interval(minutes: 60), createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, _, _) = makeEngine(reminders: [reminder], now: start)

        // Fire four times, complete two of them.
        for index in 0..<4 {
            clock.advance(by: 60 * 60)
            engine.tick()
            if index % 2 == 0 { engine.complete(id: reminder.id) }
        }

        let adherence = engine.adherence(for: reminder.id, since: start)
        XCTAssertEqual(adherence ?? 0, 0.5, accuracy: 0.001)
    }

    func testAdherenceIsNilWhenNothingFired() {
        let start = date(2026, 3, 10, 9, 0)
        let reminder = Reminder(title: "Quiet", schedule: .interval(minutes: 60))
        let (engine, _, _, _) = makeEngine(reminders: [reminder], now: start)
        XCTAssertNil(engine.adherence(for: reminder.id, since: start))
    }

    func testClearHistoryEmptiesEvents() {
        let start = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Tilt", schedule: .interval(minutes: 10), createdAt: start
        )
        reminder.lastFiredAt = start
        let (engine, clock, _, _) = makeEngine(reminders: [reminder], now: start)

        clock.advance(by: 10 * 60)
        engine.tick()
        XCTAssertFalse(engine.events.isEmpty)

        engine.clearHistory()
        XCTAssertTrue(engine.events.isEmpty)
    }

    // MARK: - Next up

    func testNextUpTracksSoonestEnabledReminder() {
        let start = date(2026, 3, 10, 9, 0)
        var hourly = Reminder(
            title: "Tilt", schedule: .interval(minutes: 60), createdAt: start
        )
        hourly.lastFiredAt = start
        var frequent = Reminder(
            title: "Shift", schedule: .interval(minutes: 20), createdAt: start
        )
        frequent.lastFiredAt = start
        let (engine, _, _, _) = makeEngine(reminders: [hourly, frequent], now: start)

        XCTAssertEqual(engine.nextUp?.reminder.title, "Shift")

        engine.setEnabled(false, for: frequent.id)
        XCTAssertEqual(engine.nextUp?.reminder.title, "Tilt")
    }
}

File: Windows/Pauselet.Core.Tests/ReminderEngineTests.cs
using NodaTime;
using Pauselet.Core;
using Xunit;

namespace Pauselet.Core.Tests;

public class ReminderEngineTests
{
    private static readonly DateTimeZone Utc = DateTimeZone.Utc;

    private static Instant Date(int year, int month, int day, int hour = 0, int minute = 0) =>
        TestDates.At(Utc, year, month, day, hour, minute);

    /// <summary>Builds an engine with a controllable clock.</summary>
    private static (ReminderEngine, MutableDateProvider, RecordingPresenter, InMemoryDataStore)
        MakeEngine(IReadOnlyList<Reminder>? reminders = null, Settings? settings = null,
                   Instant now = default)
    {
        var store = new InMemoryDataStore(new AppData
        {
            Reminders = reminders ?? [],
            Settings = settings ?? new Settings(),
            Events = [],
        });
        var clock = new MutableDateProvider(now);
        var presenter = new RecordingPresenter();
        var engine = new ReminderEngine(store, clock, presenter, Utc);
        return (engine, clock, presenter, store);
    }

    // MARK: - Firing

    [Fact]
    public void TickFiresDueReminderAndPresentsIt()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Weight Shift", Schedule = new Schedule.Interval(20),
            Priority = Priority.Subtle, CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, presenter, _) = MakeEngine([reminder], now: start);

        Assert.Empty(engine.Tick()); // Nothing is due yet
        Assert.Empty(presenter.Presented);

        clock.AdvanceSeconds(20 * 60);
        var fired = engine.Tick();
        Assert.Single(fired);
        Assert.Equal(["Weight Shift"], presenter.Presented.Select(r => r.Title).ToArray());
    }

    /// <summary>
    /// The critical bug this guards: repeated ticks in the same due window must
    /// not re-fire and spam the user.
    /// </summary>
    [Fact]
    public void RepeatedTicksDoNotRefireSameWindow()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Tilt Back", Schedule = new Schedule.Interval(60), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, presenter, _) = MakeEngine([reminder], now: start);

        clock.AdvanceSeconds(60 * 60);
        Assert.Single(engine.Tick());
        // Several more ticks a few seconds apart.
        for (var i = 0; i < 5; i++)
        {
            clock.AdvanceSeconds(5);
            Assert.Empty(engine.Tick());
        }
        Assert.Single(presenter.Presented);
    }

    [Fact]
    public void HigherPriorityPresentedLastSoItSitsInFront()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var subtle = new Reminder
        {
            Title = "Shift", Schedule = new Schedule.Interval(20),
            Priority = Priority.Subtle, CreatedAt = start,
        } with { LastFiredAt = start };
        var critical = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(20),
            Priority = Priority.Critical, CreatedAt = start,
        } with { LastFiredAt = start };

        var (engine, clock, presenter, _) = MakeEngine([subtle, critical], now: start);
        clock.AdvanceSeconds(20 * 60);
        engine.Tick();

        Assert.Equal(["Shift", "Tilt"], presenter.Presented.Select(r => r.Title).ToArray());
    }

    [Fact]
    public void DisabledReminderDoesNotFire()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Off", Schedule = new Schedule.Interval(10), CreatedAt = start,
        } with { IsEnabled = false, LastFiredAt = start };
        var (engine, clock, presenter, _) = MakeEngine([reminder], now: start);

        clock.AdvanceSeconds(60 * 60);
        Assert.Empty(engine.Tick());
        Assert.Empty(presenter.Presented);
    }

    // MARK: - Pause

    [Fact]
    public void PausedEngineFiresNothing()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(20), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, presenter, _) = MakeEngine([reminder], now: start);

        engine.SetPaused(true);
        clock.AdvanceSeconds(60 * 60);
        Assert.Empty(engine.Tick());
        Assert.Empty(presenter.Presented);
        Assert.Equal(1, presenter.DismissAllCount); // Pausing clears anything on screen
    }

    [Fact]
    public void TimedPauseExpiresAndResumesFiring()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(20), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, presenter, _) = MakeEngine([reminder], now: start);

        engine.PauseFor(30);
        clock.AdvanceSeconds(20 * 60);
        Assert.Empty(engine.Tick()); // Still paused

        clock.AdvanceSeconds(15 * 60); // 35 min total, pause has expired
        Assert.Single(engine.Tick());
        Assert.False(engine.Settings.IsPaused); // Pause auto-clears on expiry
        Assert.Null(engine.Settings.PausedUntil);
        Assert.Single(presenter.Presented);
    }

    /// <summary>Coming back from a long pause should not dump a backlog on the user.</summary>
    [Fact]
    public void ResumeReanchorsIntervalRemindersSoNoBacklogFires()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Shift", Schedule = new Schedule.Interval(20), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, presenter, _) = MakeEngine([reminder], now: start);

        engine.SetPaused(true);
        clock.AdvanceSeconds(4 * 60 * 60); // paused for 4 hours
        engine.Resume();

        Assert.Empty(engine.Tick()); // Nothing should fire the instant we resume
        clock.AdvanceSeconds(20 * 60);
        Assert.Single(engine.Tick()); // Next fire is a full interval after resume
        Assert.Single(presenter.Presented);
    }

    // MARK: - Quiet hours

    [Fact]
    public void QuietHoursSuppressNonCriticalButAllowCritical()
    {
        var night = Date(2026, 3, 10, 23, 0);
        var settings = new Settings
        {
            QuietHours = new QuietHours
            {
                IsEnabled = true, StartHour = 22, StartMinute = 0,
                EndHour = 7, EndMinute = 0, AllowsCritical = true,
            },
        };
        var normal = new Reminder
        {
            Title = "Water", Schedule = new Schedule.Interval(20),
            Priority = Priority.Normal, CreatedAt = night,
        } with { LastFiredAt = night };
        var critical = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(20),
            Priority = Priority.Critical, CreatedAt = night,
        } with { LastFiredAt = night };

        var (engine, clock, presenter, _) = MakeEngine([normal, critical], settings, night);
        clock.AdvanceSeconds(20 * 60);
        var fired = engine.Tick();

        Assert.Equal(["Tilt"], fired.Select(r => r.Title).ToArray());
        Assert.Equal(["Tilt"], presenter.Presented.Select(r => r.Title).ToArray());
    }

    // MARK: - User responses

    [Fact]
    public void CompleteResetsTheIntervalAndRecordsHistory()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(60), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, _, _) = MakeEngine([reminder], now: start);

        clock.AdvanceSeconds(60 * 60);
        engine.Tick();
        clock.AdvanceSeconds(5 * 60);
        engine.Complete(reminder.Id);

        var stored = engine.ReminderWithId(reminder.Id);
        Assert.Equal(clock.Now, stored?.LastFiredAt);
        Assert.Equal(clock.Now, stored?.LastAcknowledgedAt);
        Assert.Equal(
            [ReminderEvent.Outcome.Fired, ReminderEvent.Outcome.Completed],
            engine.Events.Select(e => e.EventOutcome).ToArray()
        );

        // The next fire is a full hour after completion, not after the original fire.
        clock.AdvanceSeconds(58 * 60);
        Assert.Empty(engine.Tick());
        clock.AdvanceSeconds(3 * 60);
        Assert.Single(engine.Tick());
    }

    [Fact]
    public void SnoozeDelaysNextFireByConfiguredMinutes()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(60), CreatedAt = start,
        } with { LastFiredAt = start };
        var settings = new Settings { SnoozeMinutes = 10 };
        var (engine, clock, presenter, _) = MakeEngine([reminder], settings, start);

        clock.AdvanceSeconds(60 * 60);
        engine.Tick();
        engine.Snooze(reminder.Id);

        clock.AdvanceSeconds(9 * 60);
        Assert.Empty(engine.Tick()); // Still snoozed
        clock.AdvanceSeconds(2 * 60);
        Assert.Single(engine.Tick()); // Fires again after the snooze
        Assert.Equal(2, presenter.Presented.Count);
    }

    [Fact]
    public void SnoozeWithExplicitMinutesOverridesSetting()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(60), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, _, _) = MakeEngine([reminder], now: start);

        clock.AdvanceSeconds(60 * 60);
        engine.Tick();
        engine.Snooze(reminder.Id, 2);

        clock.AdvanceSeconds(60);
        Assert.Empty(engine.Tick());
        clock.AdvanceSeconds(90);
        Assert.Single(engine.Tick());
    }

    [Fact]
    public void DismissRecordsHistoryWithoutResettingSchedule()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Water", Schedule = new Schedule.Interval(60), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, _, _) = MakeEngine([reminder], now: start);

        clock.AdvanceSeconds(60 * 60);
        engine.Tick();
        var firedAt = engine.ReminderWithId(reminder.Id)?.LastFiredAt;
        clock.AdvanceSeconds(60);
        engine.Dismiss(reminder.Id);

        Assert.Equal(firedAt, engine.ReminderWithId(reminder.Id)?.LastFiredAt);
        Assert.Equal(
            [ReminderEvent.Outcome.Fired, ReminderEvent.Outcome.Dismissed],
            engine.Events.Select(e => e.EventOutcome).ToArray()
        );
    }

    // MARK: - CRUD

    [Fact]
    public void AddedReminderDoesNotFireImmediately()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var (engine, clock, _, _) = MakeEngine(now: start);

        engine.Add(new Reminder
        {
            Title = "New",
            Schedule = new Schedule.Interval(15),
            CreatedAt = Date(2020, 1, 1), // deliberately stale
        });
        Assert.Empty(engine.Tick()); // Add re-anchors creation to now

        clock.AdvanceSeconds(15 * 60);
        Assert.Single(engine.Tick());
    }

    [Fact]
    public void ReenablingReminderReanchorsSoItDoesNotFireInstantly()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Shift", Schedule = new Schedule.Interval(20), CreatedAt = start,
        } with { IsEnabled = false, LastFiredAt = Date(2020, 1, 1) };
        var (engine, clock, _, _) = MakeEngine([reminder], now: start);

        engine.SetEnabled(true, reminder.Id);
        Assert.Empty(engine.Tick()); // Should not fire from a stale timestamp

        clock.AdvanceSeconds(20 * 60);
        Assert.Single(engine.Tick());
    }

    [Fact]
    public void DeleteRemovesReminder()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Gone", Schedule = new Schedule.Interval(10), CreatedAt = start,
        };
        var (engine, clock, _, _) = MakeEngine([reminder], now: start);

        engine.Delete(reminder.Id);
        Assert.Empty(engine.Reminders);
        clock.AdvanceSeconds(60 * 60);
        Assert.Empty(engine.Tick());
    }

    [Fact]
    public void UpdateChangesScheduleInPlace()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Shift", Schedule = new Schedule.Interval(60), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, _, _) = MakeEngine([reminder], now: start);

        engine.Update(reminder with { Schedule = new Schedule.Interval(10) });

        clock.AdvanceSeconds(10 * 60);
        Assert.Single(engine.Tick());
    }

    // MARK: - Persistence

    [Fact]
    public void ChangesArePersistedToStore()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var (engine, _, _, store) = MakeEngine(now: start);

        engine.Add(new Reminder { Title = "Persisted", Schedule = new Schedule.Interval(30) });

        Assert.Equal(["Persisted"], store.Data.Reminders.Select(r => r.Title).ToArray());
        Assert.True(store.SaveCount > 0);
    }

    [Fact]
    public void EngineLoadsExistingDataFromStore()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var existing = new Reminder { Title = "Loaded", Schedule = new Schedule.Interval(45) };
        var (engine, _, _, _) = MakeEngine([existing], now: start);
        Assert.Equal(["Loaded"], engine.Reminders.Select(r => r.Title).ToArray());
    }

    // MARK: - History and stats

    [Fact]
    public void HistoryIsCappedToAvoidUnboundedGrowth()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Chatty", Schedule = new Schedule.Interval(1), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, _, _) = MakeEngine([reminder], now: start);

        // Fire well past the cap.
        for (var i = 0; i < ReminderEngine.MaxStoredEvents + 50; i++)
        {
            clock.AdvanceSeconds(60);
            engine.Tick();
        }
        Assert.True(engine.Events.Count <= ReminderEngine.MaxStoredEvents);
    }

    [Fact]
    public void AdherenceReflectsCompletionRate()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(60), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, _, _) = MakeEngine([reminder], now: start);

        // Fire four times, complete two of them.
        for (var index = 0; index < 4; index++)
        {
            clock.AdvanceSeconds(60 * 60);
            engine.Tick();
            if (index % 2 == 0) engine.Complete(reminder.Id);
        }

        var adherence = engine.Adherence(reminder.Id, start);
        Assert.NotNull(adherence);
        Assert.Equal(0.5, adherence!.Value, 3);
    }

    [Fact]
    public void AdherenceIsNullWhenNothingFired()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder { Title = "Quiet", Schedule = new Schedule.Interval(60) };
        var (engine, _, _, _) = MakeEngine([reminder], now: start);
        Assert.Null(engine.Adherence(reminder.Id, start));
    }

    [Fact]
    public void ClearHistoryEmptiesEvents()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(10), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, clock, _, _) = MakeEngine([reminder], now: start);

        clock.AdvanceSeconds(10 * 60);
        engine.Tick();
        Assert.NotEmpty(engine.Events);

        engine.ClearHistory();
        Assert.Empty(engine.Events);
    }

    // MARK: - Next up

    [Fact]
    public void NextUpTracksSoonestEnabledReminder()
    {
        var start = Date(2026, 3, 10, 9, 0);
        var hourly = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(60), CreatedAt = start,
        } with { LastFiredAt = start };
        var frequent = new Reminder
        {
            Title = "Shift", Schedule = new Schedule.Interval(20), CreatedAt = start,
        } with { LastFiredAt = start };
        var (engine, _, _, _) = MakeEngine([hourly, frequent], now: start);

        Assert.Equal("Shift", engine.NextUp?.Reminder.Title);

        engine.SetEnabled(false, frequent.Id);
        Assert.Equal("Tilt", engine.NextUp?.Reminder.Title);
    }
}
